channels/OFF-CHAIN.md

```diff
@@ -1009,3 +1010,61 @@ contract update](../serializations.md#channel-off-chain-update-call-contract). I
 information needed for a contract call, including the contract address. The other participant co-signs
 the changes and the contract call is considered to be executed. Its results can be extracted from the calls tree in the state tree.
 
+### Contracts referring on-chain data
+
+Contracts being used in channels off-chain calls have the exact same semantics
+as one being used on-chain. Because of the different environment though, those
+might have different results as the on-chain ones.
+
+Since there is no signle source of truth, each participant considers one's chain
+to be the correct one. This is essential for the trustless environment. Every
+contract call is pinned to the top of the chain, as it is seen by each
+participant. This could cause some differences in local contracts executions.
+If those can not be resolved, participants can always rely on the blockchain
+as an arbiter by using forcing of progress. Then the top is being used as it
+is seen by the Bitcoin-NG leader.
+
+It is worth mentioning that both local contracts' executions and the forced
+progress ones must be fully deterministic and this implies some restrictions
+on using on-chain objects in the off-chain contracts. This is especially true
+for the `Chain` API:
+
+* `Chain.coinbase` - in off-chain calls we do not have miners and thus we can not
+  use the beneficiary of the generation. An address of `0` is returned instead.
+* `Chain.timestamp` - the time as in the latest block (either key or micro
+  block) as seen by the channel participant executing the contract call
+* `Chain.block_height` - the height of the chain as seen by the channel
+  participant executing the contract call
+* `Chain.difficulty` - the difficulty of the chain as seen by the channel
+  participant executing the contract call
+
+In off-chain calls we allow only look-ups for various on-chain objects.
+Creating new on-chain objects does not make any sense as those off-chain
+calls will not make it on-chain. The following look-ups are allowed
+* `AENS.resolve(name : string, pointer : string) : option(string)` - a name's
+  pointer data
+* `Oracle.get_answer(oracle_id : oracle(query_t, answer_t), query_id :
+  oracle_query(query_t, answer_t)) : option(answer_t)` - and oracle's answer
+  to a specific query
+* `Oracle.get_question(oracle_id : oracle(query_t, answer_t), query_id :
+  oracle_query(query_t, answer_t)) : query_t` - and oracle's query
+  question
+
+Registration and updates of names or asking of oracles is impossible in the
+off-chain contract calls.
+
+Using on-chain contracts in off-chain ones is a tricky task. On-chain
```

Comment:
@knarz please confirm this is how we want it. The following paragraphs describe how it is currently implemented.

Reply:
Pulling state into the channel versus referencing it, comes with the same caveats that you have in C, I suppose. Copying is both slow and costly while just referencing it exposes you to the risk of someone messing with your memory—in the future anyway, since the reference was valid at a given blockheight, which could be included, to make it future proof but would also increase the burden of validation.

With that in mind, I suppose it's better to just copy whatever you might need for execution.